compiler.py

```diff
@@ -0,0 +1,637 @@
+            elif argc == 2:
+                start, end, step = *map(self.as_value, argv), 1
+            elif argc == 3:
+                start, end, step = map(self.as_value, argv)
+            else:
+                raise CompilerError(ERR_BAD_ITER_ARGS, node)
+        else:
+            raise CompilerError(ERR_UNSUPPORTED_ITER, node)
+
+        self._ins.append(f'set {it} {start}')
+
+        self._ins.append(f'jump {{}} greaterThanEq {it} {end}')
+        # TODO  if the step in range(start,stop,step) is negative, the relevant test is lessThanEq
+        condition = len(self._ins) - 1
+
+        self._ins.extend(inject)
+        for subnode in node.body:
+            self.visit(subnode)
+
+        self._ins.append(f'op add {it} {it} {step}')
+        self._ins.append(f'jump {condition} always')
+        self._ins[condition] = self._ins[condition].format(len(self._ins))
+
+    def visit_FunctionDef(self, node):
+        """This will be called for any* function definition of the form def fname( args ): body """
+
+        # TODO forbid recursion (or implement it by storing and restoring everything from stack)
+        #      Telos: There's no need to actively forbid recursion, though it may be good to warn about dangers of it.
+        #             So long as there is no stack to overflow, recursion can work perfectly fine as a form of infinite
+        #             loop, or the recursion can be escaped with an end() command, a jump_to( label ) command, or simply
+        #             by entering into some other infinite loop and never return-ing. The key point is just to warn that
+        #             recursive calls will overwrite the value of the return pointer (if we follow my suggestion of
+        #             implementing that in a dummy variable to forego needing a stack) and thereby lose track of
+        #             whatever it was that called the function in the first place.  If you instead push return addresses
+        #             onto a stack, and wait to pop them at return-time, then recursion will still be fine, and will
+        #             even return all the way back to the first caller, with the only caveat being that all variables
+        #             are global, effectively including function arguments, so you have to be a bit cautious how
+        #             you write the code.
+        # TODO local variable namespace per-function
+        #      Telos:  I doubt that mindustry's processor-speed is fast enough to really make this worthwhile.
+        #              I'd recommend instead just documenting that mlog uses a single global namespace, and that
+        #              each call to a function overwrites the prior global values of its arguments.
+
+        if self._in_def is not None:
+            raise CompilerError(ERR_NESTED_DEF, node)
+        # TODO there's no reason we *couldn't* allow nested function defns, though it may be more tricky than it's worth

```

Comment:
They *can* be supported, but it's extremely challenging due to the fact they capture their environment.